hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/HoodieCompactor.java

```diff
@@ -247,6 +256,17 @@ public List<WriteStatus> compact(HoodieCompactionHandler compactionHandler,
     }).collect(toList());
   }
 
+  private InstantRange getMetadataLogReaderInstantRange(HoodieTableMetaClient metadataMetaClient, String metadataBasePath) {
```

Comment:
@danny0405 @nsivabalan @prashantwason I'm a little rusty on the interplay between MDT compaction and inflight instants on data table's timeline.  While I'm reading more code to refresh my memory, could you folks remind me of how the the inflight instants are handled on data table's timeline after MDT compaction is done in this case?  I'm trying to see if the changes in this PR is safe.

Suppose there's an inflight instant before a complete instant in the data table's timeline (`C1.commit, C2.commit.inflight, C3.commit`) due to inflight table service or concurrent writers.  There's a hole in the timeline.  In MDT timeline, C1, C2, C3 are all committed and generate log files, and C2 should be excluded in the metadata table merged log scanner, since C2 is not committed in the table.  If MDT compaction happens at this point, C2 is skipped in compaction.  Is there any case where C2's log file in MDT is kept for later and it will be missed in MDT, since it's in the old file slice?

Reply:
>  Is there any case where C2's log file in MDT is kept for later and it will be missed in MDT, since it's in the old file slice?

Good question, the log would then be removed by the MDT cleaner ? So it would not be kept forever. When new rollback commit for C2 from data table is committed into MDT again, there could be some tumestone records there which would be ignored (by default allowed by master codebase).

Reply:
@yihua There are these case for C2.commit.inflight to be there:
1. C2 is still in inflight and not committed to MDT yet
2. C2 committed successfully to MDT and still in process of completing the commit on MDT (C2's process not exited yet)
3. C2 committed successfully to MDT but failed to complete  (C2's process failed)

In the multi-writer scenario, the compaction can the initiated by the process which created C3 and that may lead to the case 2 above. In this case leaving C2 out in compaction would be an error as it is going to be a completed commit on dataset. 

This is the reason why we have checks that we will initiate MDT compaction if there are pending inflights on the dataset timeline.